controllers/watcherapi_controller.go

```diff
@@ -552,61 +550,110 @@ func (r *WatcherAPIReconciler) ensureServiceExposed(
 
-	for endpointType := range instance.Spec.Override.Service {
+	for endpointType, data := range ports {
+		endpointTypeStr := string(endpointType)
+		endpointName := watcher.ServiceName + "-" + endpointTypeStr
 		svcOverride := instance.Spec.Override.Service[endpointType]
-		portCfg := ports[endpointType]
-		portCfg.MetalLB = &endpoint.MetalLBData{
-			IPAddressPool:   svcOverride.IPAddressPool,
-			SharedIP:        svcOverride.SharedIP,
-			SharedIPKey:     svcOverride.SharedIPKey,
-			LoadBalancerIPs: svcOverride.LoadBalancerIPs,
+		if svcOverride.EmbeddedLabelsAnnotations == nil {
+			svcOverride.EmbeddedLabelsAnnotations = &service.EmbeddedLabelsAnnotations{}
 		}
 
-		ports[endpointType] = portCfg
-	}
+		exportLabels := util.MergeStringMaps(
+			getAPIServiceLabels(),
+			map[string]string{
+				service.AnnotationEndpointKey: endpointTypeStr,
+			},
+		)
 
-	apiEndpoints, ctrlResult, err := endpoint.ExposeEndpoints(
-		ctx,
-		helper,
-		watcher.ServiceName,
-		getAPIServiceLabels(),
-		ports,
-		r.RequeueTimeout,
-	)
-	if err != nil {
-		instance.Status.Conditions.Set(condition.FalseCondition(
-			condition.ExposeServiceReadyCondition,
-			condition.ErrorReason,
-			condition.SeverityWarning,
-			condition.ExposeServiceReadyErrorMessage,
-			err.Error(),
-		))
-		return nil, ctrlResult, err
-	} else if (ctrlResult != ctrl.Result{}) {
-		instance.Status.Conditions.Set(condition.FalseCondition(
-			condition.ExposeServiceReadyCondition,
-			condition.RequestedReason,
-			condition.SeverityInfo,
-			condition.ExposeServiceReadyRunningMessage,
-		))
-		return nil, ctrlResult, nil
-	}
-
-	// fix wrongly formatted endpoint url gotten from the lib-common
-	// ExposeEndpoints function
-	for endpointType, endpointURL := range apiEndpoints {
-		// fix repeated '://' in endpoint url
-		endpointURL = strings.Replace(endpointURL, "://://", "://", 1)
-		if endpointType == string(service.EndpointPublic) {
-			// remove trailing port number
-			url, _ := url.Parse(endpointURL)
-			endpointURL = fmt.Sprintf("%s://%s", url.Scheme, url.Hostname())
+		// create the service
+		svc, err := service.NewService(
+			service.GenericService(&service.GenericServiceDetails{
+				Name:      endpointName,
+				Namespace: instance.Namespace,
+				Labels:    exportLabels,
+				Selector:  getAPIServiceLabels(),
+				Port: service.GenericServicePort{
+					Name:     endpointName,
+					Port:     data.Port,
+					Protocol: corev1.ProtocolTCP,
+				},
+			}),
+			5,
+			&svcOverride.OverrideSpec,
+		)
+		if err != nil {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				condition.CreateServiceReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				condition.CreateServiceReadyErrorMessage,
+				err.Error(),
+			))
+			return nil, ctrl.Result{}, err
+		}
+
+		svc.AddAnnotation(map[string]string{
+			service.AnnotationEndpointKey: endpointTypeStr,
+		})
+
+		// add Annotation to whether creating an ingress is required or not
```

Comment:
we need to remmeber to remove this once we integrate with the OpenStack operator but ya this is what we discussed doing until then